code/components/citizen-server-impl/src/state/ServerSetters.cpp

```diff
@@ -447,6 +539,163 @@ static InitFunction initFunction([]()
 			ctx.SetResult(sgs->MakeScriptHandle(entity));
 		});
 
+		fx::ScriptEngine::RegisterNativeHandler("CREATE_TRAIN", [=](fx::ScriptContext& ctx)
+		{
+			uint32_t resourceHash = 0;
+
+			fx::OMPtr<IScriptRuntime> runtime;
+
+			if (FX_SUCCEEDED(fx::GetCurrentScriptRuntime(&runtime)))
+			{
+				fx::Resource* resource = reinterpret_cast<fx::Resource*>(runtime->GetParentObject());
+
+				if (resource)
+				{
+					resourceHash = HashString(resource->GetName().c_str());
+				}
+			}
+
+			uint32_t modelHash = ctx.GetArgument<uint32_t>(0);
+			float x = ctx.GetArgument<float>(1);
+			float y = ctx.GetArgument<float>(2);
+			float z = ctx.GetArgument<float>(3);
+			bool direction = ctx.GetArgument<bool>(4);
+			bool stopAtStations = ctx.GetArgument<bool>(5);
+			float speed = ctx.GetArgument<float>(6);
+			uint32_t trackId = ctx.GetArgument<uint32_t>(7);
+			uint32_t trainConfigIndex = ctx.CheckArgument<uint32_t>(8);
+
+			//Don't create train if an invalid trackid has been provided.
+			if (trackId != 0 && trackId != 3)
+			{
+				throw std::runtime_error(va("Tried to spawn train on invalid track id: %u", trackId));
+
+				ctx.SetResult(0);
+				return;
+			}
+
+			//Don't create train if an invalid trainConfigIndex has been provided, otherwises the client crashes.
+			if (trainConfigIndex > (uint32_t)(Is2802() ? 26 : (Is2372() ? 25 : 24)))
```

Comment:
The same applies for train config IDs, as a server can add new ones to the game. This check stops those from being usable.